File: go/attribution/validate.go
package attribution

import (
	"bytes"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
)

type ValidationIssue struct {
	Path    string `json:"path"`
	Message string `json:"message"`
}

type ValidationResult struct {
	Success bool              `json:"success"`
	Issues  []ValidationIssue `json:"issues"`
}

var (
	uuidPattern      = regexp.MustCompile(`(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)
	canonicalPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9._-]{0,127}$`)
	positiveDecimal  = regexp.MustCompile(`^(?:0|[1-9][0-9]{0,25})(?:\.[0-9]{1,12})?$`)
	currencyPattern  = regexp.MustCompile(`^[A-Z]{3}$`)
	timestampPattern = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?(?:Z|[+-]\d{2}:\d{2})$`)
)

var components = stringSet("llm", "telephony", "voice_platform", "speech_to_text", "text_to_speech", "realtime_transport", "recording", "post_call_analysis", "compute", "gpu", "network", "storage", "external")
var resourceTypes = stringSet("model", "sku", "instance", "endpoint", "session", "other")
var lifecycleStates = stringSet("pending", "provisional", "final", "voided")
var evidenceSources = stringSet("provider_reported", "sdk_catalog", "sdk_rate_registry", "manual")
var confidences = stringSet("exact", "computed", "estimated", "unknown")

func stringSet(values ...string) map[string]bool {
	set := make(map[string]bool, len(values))
	for _, value := range values {
		set[value] = true
	}
	return set
}

func ValidateEventV2(value interface{}) ValidationResult {
	var event map[string]interface{}
	b, err := json.Marshal(value)
	if err == nil {
		decoder := json.NewDecoder(bytes.NewReader(b))
		decoder.UseNumber()
		err = decoder.Decode(&event)
	}
	issues := make([]ValidationIssue, 0)
	add := func(path, message string) { issues = append(issues, ValidationIssue{Path: path, Message: message}) }
	if err != nil || event == nil {
		add("", "Event must be an object")
		return ValidationResult{Success: false, Issues: issues}
	}
	unknownKeys(event, stringSet("schema_version", "event_id", "task_id", "occurred_at", "observed_at", "component", "provider", "resource", "lifecycle", "usage_period", "usage", "cost_evidence", "retry_of"), "", add)
	if event["schema_version"] != "2" {
		add("schema_version", "Must equal 2")
	}
	validString(event["event_id"], "event_id", uuidPattern, add)
	validString(event["task_id"], "task_id", uuidPattern, add)
	parseTimestamp(event["occurred_at"], "occurred_at", add)
	parseTimestamp(event["observed_at"], "observed_at", add)
	component, ok := event["component"].(string)
	if !ok || !components[component] {
		add("component", "Unknown attribution component")
	}
	if retry, exists := event["retry_of"]; exists {
		validString(retry, "retry_of", uuidPattern, add)
	}

	provider, ok := event["provider"].(map[string]interface{})
	if !ok {
		add("provider", "Provider must be an object")
	} else {
		unknownKeys(provider, stringSet("name", "service", "record_id", "region"), "provider", add)
		validString(provider["name"], "provider.name", canonicalPattern, add)
		validString(provider["service"], "provider.service", canonicalPattern, add)
		if record, exists := provider["record_id"]; exists {
			s, ok := record.(string)
			if !ok || len(s) < 1 || len(s) > 256 {
				add("provider.record_id", "Invalid provider record ID")
			}
		}
		if region, exists := provider["region"]; exists {
			validString(region, "provider.region", canonicalPattern, add)
		}
	}

	if raw, exists := event["resource"]; exists {
		resource, ok := raw.(map[string]interface{})
		if !ok {
			add("resource", "Resource must be an object")
		} else {
			unknownKeys(resource, stringSet("type", "id"), "resource", add)
			t, ok := resource["type"].(string)
			if !ok || !resourceTypes[t] {
				add("resource.type", "Invalid resource type")
			}
			id, ok := resource["id"].(string)
			if !ok || len(id) < 1 || len(id) > 256 {
				add("resource.id", "Invalid resource ID")
			}
		}
	}

	state := ""
	revision := int64(0)
	lifecycle, ok := event["lifecycle"].(map[string]interface{})
	if !ok {
		add("lifecycle", "Lifecycle must be an object")
	} else {
		unknownKeys(lifecycle, stringSet("state", "revision"), "lifecycle", add)
		state, ok = lifecycle["state"].(string)
		if !ok || !lifecycleStates[state] {
			add("lifecycle.state", "Invalid lifecycle state")
			state = ""
		}
		n, ok := lifecycle["revision"].(json.Number)
		if !ok {
			add("lifecycle.revision", "Revision must be a positive integer")
		} else if revision, err = n.Int64(); err != nil || revision < 1 || revision > 2147483647 {
			add("lifecycle.revision", "Revision must be a positive integer")
			revision = 0
		}
	}

	var period map[string]interface{}
	var start, end time.Time
	var startOK, endOK bool
	if raw, exists := event["usage_period"]; exists {
		period, ok = raw.(map[string]interface{})
		if !ok {
			add("usage_period", "Usage period must be an object")
		} else {
			unknownKeys(period, stringSet("start_at", "end_at"), "usage_period", add)
			start, startOK = parseTimestamp(period["start_at"], "usage_period.start_at", add)
			if rawEnd, exists := period["end_at"]; exists {
				end, endOK = parseTimestamp(rawEnd, "usage_period.end_at", add)
			}
			if startOK && endOK && end.Before(start) {
				add("usage_period.end_at", "End cannot precede start")
			}
		}
	}

	usage, usageOK := event["usage"].([]interface{})
	hasTime := false
	seen := map[string]bool{}
	if !usageOK {
		add("usage", "Usage must be an array")
	} else {
		if len(usage) > 32 {
			add("usage", "At most 32 usage lines are allowed")
		}
		for i, raw := range usage {
			prefix := fmt.Sprintf("usage.%d", i)
			line, ok := raw.(map[string]interface{})
			if !ok {
				add(prefix, "Usage line must be an object")
				continue
			}
			unknownKeys(line, stringSet("metric", "quantity", "unit"), prefix, add)
			metric, metricOK := line["metric"].(string)
			canonicalUnit, knownMetric := UnitByMetric[UsageMetric(metric)]
			if !metricOK || !knownMetric {
				add(prefix+".metric", "Invalid usage metric")
			} else if seen[metric] {
				add(prefix+".metric", "Duplicate usage metric")
			} else {
				seen[metric] = true
			}
			quantity, quantityOK := line["quantity"].(string)
			if !quantityOK || !positiveDecimal.MatchString(quantity) || !strings.ContainsAny(quantity, "123456789") {
				add(prefix+".quantity", "Must be a positive plain decimal string")
			}
			unit, unitOK := line["unit"].(string)
			if !unitOK || !unitKnown(unit) {
				add(prefix+".unit", "Invalid usage unit")
			} else {
				hasTime = hasTime || strings.HasSuffix(unit, "Seconds")
				if knownMetric && UsageUnit(unit) != canonicalUnit {
					add(prefix+".unit", "Metric must use its canonical unit")
				}
			}
		}
	}

	var cost map[string]interface{}
	if raw, exists := event["cost_evidence"]; exists {
		cost, ok = raw.(map[string]interface{})
		if !ok {
			add("cost_evidence", "Cost evidence must be an object")
		} else {
			unknownKeys(cost, stringSet("amount", "currency", "source", "confidence", "pricing_version"), "cost_evidence", add)
			amount, amountOK := cost["amount"].(string)
			if !amountOK || !positiveDecimal.MatchString(amount) || !strings.ContainsAny(amount, "123456789") {
				add("cost_evidence.amount", "Must be a positive plain decimal string")
			}
			currency, currencyOK := cost["currency"].(string)
			if !currencyOK || !currencyPattern.MatchString(currency) {
				add("cost_evidence.currency", "Invalid currency")
			}
			source, sourceOK := cost["source"].(string)
			if !sourceOK || !evidenceSources[source] {
				add("cost_evidence.source", "Invalid evidence source")
				source = ""
			}
			confidence, confidenceOK := cost["confidence"].(string)
			if !confidenceOK || !confidences[confidence] {
				add("cost_evidence.confidence", "Invalid confidence")
				confidence = ""
			}
			if pv, exists := cost["pricing_version"]; exists {
				s, ok := pv.(string)
				if !ok || len(s) < 1 || len(s) > 128 {
					add("cost_evidence.pricing_version", "Invalid pricing version")
				}
			}
			if source == "provider_reported" && confidence != "" && confidence != "exact" && confidence != "estimated" {
				add("cost_evidence.confidence", "Provider-reported cost must be exact or estimated")
			}
			if (source == "sdk_catalog" || source == "sdk_rate_registry") && confidence == "exact" {
				add("cost_evidence.confidence", "SDK-derived cost cannot be exact")
			}
			if source == "sdk_catalog" || source == "sdk_rate_registry" {
				if _, exists := cost["pricing_version"]; !exists {
					add("cost_evidence.pricing_version", "SDK-derived cost requires pricing_version")
				}
			}
		}
	}

	if hasTime && (state == "provisional" || state == "final") && !endOK {
		add("usage_period.end_at", "Finalized time-based usage requires a closed usage period")
	}
	usageLen := 0
	if usageOK {
		usageLen = len(usage)
	}
	switch state {
	case "pending":
		if usageLen != 0 {
			add("usage", "Pending events cannot assert usage")
		}
		if _, exists := event["cost_evidence"]; exists {
			add("cost_evidence", "Pending events cannot assert cost")
		}
		if period != nil {
			if _, exists := period["end_at"]; exists {
				add("usage_period.end_at", "Pending events cannot close usage")
			}
		}
	case "provisional":
		if usageLen == 0 {
			add("usage", "Provisional events require usage")
		}
		if cost != nil && cost["confidence"] == "exact" {
			add("cost_evidence.confidence", "Provisional cost cannot be exact")
		}
	case "final":
		if usageLen == 0 {
			add("usage", "Final events require usage")
		}
	case "voided":
		if revision == 1 {
			add("lifecycle.revision", "Voided events must supersede an earlier revision")
		}
		if usageLen != 0 {
			add("usage", "Voided events must be tombstones")
		} else if _, exists := event["cost_evidence"]; exists {
			add("usage", "Voided events must be tombstones")
		}
	}
	return ValidationResult{Success: len(issues) == 0, Issues: issues}
}

func unknownKeys(value map[string]interface{}, allowed map[string]bool, prefix string, add func(string, string)) {
	for key := range value {
		if !allowed[key] {
			path := key
			if prefix != "" {
				path = prefix + "." + key
			}
			add(path, "Unknown field")
		}
	}
}

func validString(value interface{}, path string, pattern *regexp.Regexp, add func(string, string)) bool {
	s, ok := value.(string)
	if !ok || s == "" || (pattern != nil && !pattern.MatchString(s)) {
		add(path, "Invalid string value")
		return false
	}
	return true
}

func parseTimestamp(value interface{}, path string, add func(string, string)) (time.Time, bool) {
	s, ok := value.(string)
	if !ok || !timestampPattern.MatchString(s) {
		add(path, "Invalid string value")
		return time.Time{}, false
	}
	parsed, err := time.Parse(time.RFC3339Nano, s)
	if err != nil {
		add(path, "Timestamp must be a valid ISO 8601 calendar instant")
		return time.Time{}, false
	}
	return parsed.UTC(), true
}

func unitKnown(unit string) bool {
	for _, candidate := range UnitByMetric {
		if string(candidate) == unit {
			return true
		}
	}
	return false
}

File: rust/src/attribution/validate.rs
use std::collections::HashSet;
use std::sync::LazyLock;

use chrono::{DateTime, FixedOffset};
use regex::Regex;
use serde_json::{Map, Value};

use super::types::{AttributionValidationIssue, AttributionValidationResult};

static UUID_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
        .expect("valid UUID regex")
});
static CANONICAL_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[a-z0-9][a-z0-9._-]{0,127}$").expect("valid canonical-name regex")
});
static POSITIVE_DECIMAL: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^(?:0|[1-9][0-9]{0,25})(?:\.[0-9]{1,12})?$").expect("valid decimal regex")
});
static CURRENCY_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[A-Z]{3}$").expect("valid currency regex"));
static TIMESTAMP_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?(?:Z|[+-]\d{2}:\d{2})$")
        .expect("valid timestamp regex")
});

const COMPONENTS: &[&str] = &[
    "llm",
    "telephony",
    "voice_platform",
    "speech_to_text",
    "text_to_speech",
    "realtime_transport",
    "recording",
    "post_call_analysis",
    "compute",
    "gpu",
    "network",
    "storage",
    "external",
];
const RESOURCE_TYPES: &[&str] = &["model", "sku", "instance", "endpoint", "session", "other"];
const LIFECYCLE_STATES: &[&str] = &["pending", "provisional", "final", "voided"];
const EVIDENCE_SOURCES: &[&str] = &[
    "provider_reported",
    "sdk_catalog",
    "sdk_rate_registry",
    "manual",
];
const CONFIDENCES: &[&str] = &["exact", "computed", "estimated", "unknown"];

type ParsedInstant = DateTime<FixedOffset>;

pub fn validate_attribution_event_v2(value: &Value) -> AttributionValidationResult {
    let mut issues = Vec::new();
    let Some(event) = value.as_object() else {
        issue(&mut issues, "", "Event must be an object");
        return result(issues);
    };

    unknown_keys(
        event,
        &[
            "schema_version",
            "event_id",
            "task_id",
            "occurred_at",
            "observed_at",
            "component",
            "provider",
            "resource",
            "lifecycle",
            "usage_period",
            "usage",
            "cost_evidence",
            "retry_of",
        ],
        "",
        &mut issues,
    );

    if event.get("schema_version").and_then(Value::as_str) != Some("2") {
        issue(&mut issues, "schema_version", "Must equal 2");
    }
    valid_string(
        event.get("event_id"),
        "event_id",
        &UUID_PATTERN,
        &mut issues,
    );
    valid_string(event.get("task_id"), "task_id", &UUID_PATTERN, &mut issues);
    parse_timestamp(event.get("occurred_at"), "occurred_at", &mut issues);
    parse_timestamp(event.get("observed_at"), "observed_at", &mut issues);

    match event.get("component").and_then(Value::as_str) {
        Some(component) if COMPONENTS.contains(&component) => {}
        _ => issue(&mut issues, "component", "Unknown attribution component"),
    }
    if let Some(retry_of) = event.get("retry_of") {
        valid_string(Some(retry_of), "retry_of", &UUID_PATTERN, &mut issues);
    }

    validate_provider(event.get("provider"), &mut issues);
    if let Some(resource) = event.get("resource") {
        validate_resource(resource, &mut issues);
    }

    let (state, revision) = validate_lifecycle(event.get("lifecycle"), &mut issues);
    let (period_present, period_closed) =
        validate_usage_period(event.get("usage_period"), &mut issues);
    let (usage_len, has_time_usage) = validate_usage(event.get("usage"), &mut issues);
    let cost = validate_cost_evidence(event.get("cost_evidence"), &mut issues);

    if has_time_usage && matches!(state, Some("provisional" | "final")) && !period_closed {
        issue(
            &mut issues,
            "usage_period.end_at",
            "Finalized time-based usage requires a closed usage period",
        );
    }

    match state {
        Some("pending") => {
            if usage_len != Some(0) {
                issue(&mut issues, "usage", "Pending events cannot assert usage");
            }
            if event.contains_key("cost_evidence") {
                issue(
                    &mut issues,
                    "cost_evidence",
                    "Pending events cannot assert cost",
                );
            }
            if period_present && period_closed {
                issue(
                    &mut issues,
                    "usage_period.end_at",
                    "Pending events cannot close usage",
                );
            }
        }
        Some("provisional") => {
            if usage_len == Some(0) {
                issue(&mut issues, "usage", "Provisional events require usage");
            }
            if cost.confidence == Some("exact") {
                issue(
                    &mut issues,
                    "cost_evidence.confidence",
                    "Provisional cost cannot be exact",
                );
            }
        }
        Some("final") => {
            if usage_len == Some(0) {
                issue(&mut issues, "usage", "Final events require usage");
            }
        }
        Some("voided") => {
            if revision == Some(1) {
                issue(
                    &mut issues,
                    "lifecycle.revision",
                    "Voided events must supersede an earlier revision",
                );
            }
            if usage_len != Some(0) || event.contains_key("cost_evidence") {
                issue(&mut issues, "usage", "Voided events must be tombstones");
            }
        }
        _ => {}
    }

    result(issues)
}

fn validate_provider(value: Option<&Value>, issues: &mut Vec<AttributionValidationIssue>) {
    let Some(provider) = value.and_then(Value::as_object) else {
        issue(issues, "provider", "Provider must be an object");
        return;
    };
    unknown_keys(
        provider,
        &["name", "service", "record_id", "region"],
        "provider",
        issues,
    );
    valid_string(
        provider.get("name"),
        "provider.name",
        &CANONICAL_PATTERN,
        issues,
    );
    valid_string(
        provider.get("service"),
        "provider.service",
        &CANONICAL_PATTERN,
        issues,
    );
    if let Some(record_id) = provider.get("record_id") {
        match record_id.as_str() {
            Some(value) if !value.is_empty() && value.len() <= 256 => {}
            _ => issue(issues, "provider.record_id", "Invalid provider record ID"),
        }
    }
    if let Some(region) = provider.get("region") {
        valid_string(Some(region), "provider.region", &CANONICAL_PATTERN, issues);
    }
}

fn validate_resource(value: &Value, issues: &mut Vec<AttributionValidationIssue>) {
    let Some(resource) = value.as_object() else {
        issue(issues, "resource", "Resource must be an object");
        return;
    };
    unknown_keys(resource, &["type", "id"], "resource", issues);
    match resource.get("type").and_then(Value::as_str) {
        Some(resource_type) if RESOURCE_TYPES.contains(&resource_type) => {}
        _ => issue(issues, "resource.type", "Invalid resource type"),
    }
    match resource.get("id").and_then(Value::as_str) {
        Some(id) if !id.is_empty() && id.len() <= 256 => {}
        _ => issue(issues, "resource.id", "Invalid resource ID"),
    }
}

fn validate_lifecycle<'a>(
    value: Option<&'a Value>,
    issues: &mut Vec<AttributionValidationIssue>,
) -> (Option<&'a str>, Option<u64>) {
    let Some(lifecycle) = value.and_then(Value::as_object) else {
        issue(issues, "lifecycle", "Lifecycle must be an object");
        return (None, None);
    };
    unknown_keys(lifecycle, &["state", "revision"], "lifecycle", issues);
    let state = match lifecycle.get("state").and_then(Value::as_str) {
        Some(state) if LIFECYCLE_STATES.contains(&state) => Some(state),
        _ => {
            issue(issues, "lifecycle.state", "Invalid lifecycle state");
            None
        }
    };
    let revision = match lifecycle.get("revision").and_then(Value::as_u64) {
        Some(revision) if (1..=i32::MAX as u64).contains(&revision) => Some(revision),
        _ => {
            issue(
                issues,
                "lifecycle.revision",
                "Revision must be a positive integer",
            );
            None
        }
    };
    (state, revision)
}

fn validate_usage_period(
    value: Option<&Value>,
    issues: &mut Vec<AttributionValidationIssue>,
) -> (bool, bool) {
    let Some(value) = value else {
        return (false, false);
    };
    let Some(period) = value.as_object() else {
        issue(issues, "usage_period", "Usage period must be an object");
        return (true, false);
    };
    unknown_keys(period, &["start_at", "end_at"], "usage_period", issues);
    let start = parse_timestamp(period.get("start_at"), "usage_period.start_at", issues);
    let end_present = period.contains_key("end_at");
    let end = if end_present {
        parse_timestamp(period.get("end_at"), "usage_period.end_at", issues)
    } else {
        None
    };
    if let (Some(start), Some(end)) = (start, end) {
        if end < start {
            issue(issues, "usage_period.end_at", "End cannot precede start");
        }
    }
    (true, end_present)
}

fn validate_usage(
    value: Option<&Value>,
    issues: &mut Vec<AttributionValidationIssue>,
) -> (Option<usize>, bool) {
    let Some(usage) = value.and_then(Value::as_array) else {
        issue(issues, "usage", "Usage must be an array");
        return (None, false);
    };
    if usage.len() > 32 {
        issue(issues, "usage", "At most 32 usage lines are allowed");
    }
    let mut seen = HashSet::new();
    let mut has_time = false;
    for (index, raw) in usage.iter().enumerate() {
        let prefix = format!("usage.{index}");
        let Some(line) = raw.as_object() else {
            issue(issues, &prefix, "Usage line must be an object");
            continue;
        };
        unknown_keys(line, &["metric", "quantity", "unit"], &prefix, issues);
        let metric = line.get("metric").and_then(Value::as_str);
        let canonical_unit = metric.and_then(canonical_unit_for_metric);
        match (metric, canonical_unit) {
            (Some(metric), Some(_)) if !seen.insert(metric) => issue(
                issues,
                &format!("{prefix}.metric"),
                "Duplicate usage metric",
            ),
            (Some(_), Some(_)) => {}
            _ => issue(issues, &format!("{prefix}.metric"), "Invalid usage metric"),
        }
        match line.get("quantity").and_then(Value::as_str) {
            Some(quantity)
                if POSITIVE_DECIMAL.is_match(quantity)
                    && quantity.bytes().any(|byte| matches!(byte, b'1'..=b'9')) => {}
            _ => issue(
                issues,
                &format!("{prefix}.quantity"),
                "Must be a positive plain decimal string",
            ),
        }
        let unit = line.get("unit").and_then(Value::as_str);
        if !matches!(
            unit,
            Some(
                "Tokens"
                    | "Characters"
                    | "Seconds"
                    | "vCPU-Seconds"
                    | "GiB-Seconds"
                    | "GPU-Seconds"
                    | "Requests"
                    | "Calls"
                    | "Bytes"
                    | "Images"
                    | "Pages"
                    | "Credits"
            )
        ) {
            issue(issues, &format!("{prefix}.unit"), "Invalid usage unit");
        } else {
            has_time |= unit.is_some_and(|unit| unit.ends_with("Seconds"));
            if let (Some(canonical), Some(actual)) = (canonical_unit, unit) {
                if actual != canonical {
                    issue(
                        issues,
                        &format!("{prefix}.unit"),
                        "Metric must use its canonical unit",
                    );
                }
            }
        }
    }
    (Some(usage.len()), has_time)
}

#[derive(Default)]
struct ParsedCostEvidence<'a> {
    confidence: Option<&'a str>,
}

fn validate_cost_evidence<'a>(
    value: Option<&'a Value>,
    issues: &mut Vec<AttributionValidationIssue>,
) -> ParsedCostEvidence<'a> {
    let Some(value) = value else {
        return ParsedCostEvidence::default();
    };
    let Some(cost) = value.as_object() else {
        issue(issues, "cost_evidence", "Cost evidence must be an object");
        return ParsedCostEvidence::default();
    };
    unknown_keys(
        cost,
        &[
            "amount",
            "currency",
            "source",
            "confidence",
            "pricing_version",
        ],
        "cost_evidence",
        issues,
    );
    match cost.get("amount").and_then(Value::as_str) {
        Some(amount)
            if POSITIVE_DECIMAL.is_match(amount)
                && amount.bytes().any(|byte| matches!(byte, b'1'..=b'9')) => {}
        _ => issue(
            issues,
            "cost_evidence.amount",
            "Must be a positive plain decimal string",
        ),
    }
    match cost.get("currency").and_then(Value::as_str) {
        Some(currency) if CURRENCY_PATTERN.is_match(currency) => {}
        _ => issue(issues, "cost_evidence.currency", "Invalid currency"),
    }
    let source = match cost.get("source").and_then(Value::as_str) {
        Some(source) if EVIDENCE_SOURCES.contains(&source) => Some(source),
        _ => {
            issue(issues, "cost_evidence.source", "Invalid evidence source");
            None
        }
    };
    let confidence = match cost.get("confidence").and_then(Value::as_str) {
        Some(confidence) if CONFIDENCES.contains(&confidence) => Some(confidence),
        _ => {
            issue(issues, "cost_evidence.confidence", "Invalid confidence");
            None
        }
    };
    if let Some(version) = cost.get("pricing_version") {
        match version.as_str() {
            Some(version) if !version.is_empty() && version.len() <= 128 => {}
            _ => issue(
                issues,
                "cost_evidence.pricing_version",
                "Invalid pricing version",
            ),
        }
    }
    if source == Some("provider_reported")
        && confidence.is_some()
        && !matches!(confidence, Some("exact" | "estimated"))
    {
        issue(
            issues,
            "cost_evidence.confidence",
            "Provider-reported cost must be exact or estimated",
        );
    }
    if matches!(source, Some("sdk_catalog" | "sdk_rate_registry")) {
        if confidence == Some("exact") {
            issue(
                issues,
                "cost_evidence.confidence",
                "SDK-derived cost cannot be exact",
            );
        }
        if !cost.contains_key("pricing_version") {
            issue(
                issues,
                "cost_evidence.pricing_version",
                "SDK-derived cost requires pricing_version",
            );
        }
    }
    ParsedCostEvidence { confidence }
}

fn canonical_unit_for_metric(metric: &str) -> Option<&'static str> {
    match metric {
        "input_tokens"
        | "output_tokens"
        | "cache_read_input_tokens"
        | "cache_write_input_tokens"
        | "reasoning_output_tokens" => Some("Tokens"),
        "characters" => Some("Characters"),
        "audio_seconds" | "connected_seconds" | "recording_seconds" | "agent_seconds"
        | "compute_seconds" => Some("Seconds"),
        "vcpu_seconds" => Some("vCPU-Seconds"),
        "memory_gib_seconds" => Some("GiB-Seconds"),
        "gpu_seconds" => Some("GPU-Seconds"),
        "request_count" => Some("Requests"),
        "call_count" => Some("Calls"),
        "bytes_in" | "bytes_out" => Some("Bytes"),
        "image_count" => Some("Images"),
        "page_count" => Some("Pages"),
        "credit_count" => Some("Credits"),
        _ => None,
    }
}

fn parse_timestamp(
    value: Option<&Value>,
    path: &str,
    issues: &mut Vec<AttributionValidationIssue>,
) -> Option<ParsedInstant> {
    let Some(raw) = value.and_then(Value::as_str) else {
        issue(issues, path, "Invalid string value");
        return None;
    };
    if !TIMESTAMP_PATTERN.is_match(raw) {
        issue(issues, path, "Invalid string value");
        return None;
    }
    match DateTime::parse_from_rfc3339(raw) {
        Ok(parsed) => Some(parsed),
        Err(_) => {
            issue(
                issues,
                path,
                "Timestamp must be a valid ISO 8601 calendar instant",
            );
            None
        }
    }
}

fn valid_string(
    value: Option<&Value>,
    path: &str,
    pattern: &Regex,
    issues: &mut Vec<AttributionValidationIssue>,
) -> bool {
    match value.and_then(Value::as_str) {
        Some(value) if !value.is_empty() && pattern.is_match(value) => true,
        _ => {
            issue(issues, path, "Invalid string value");
            false
        }
    }
}

fn unknown_keys(
    object: &Map<String, Value>,
    allowed: &[&str],
    prefix: &str,
    issues: &mut Vec<AttributionValidationIssue>,
) {
    for key in object.keys() {
        if !allowed.contains(&key.as_str()) {
            let path = if prefix.is_empty() {
                key.clone()
            } else {
                format!("{prefix}.{key}")
            };
            issue(issues, &path, "Unknown field");
        }
    }
}

fn issue(issues: &mut Vec<AttributionValidationIssue>, path: &str, message: &str) {
    issues.push(AttributionValidationIssue {
        path: path.to_string(),
        message: message.to_string(),
    });
}

fn result(issues: Vec<AttributionValidationIssue>) -> AttributionValidationResult {
    AttributionValidationResult {
        success: issues.is_empty(),
        issues,
    }
}
